internal/impl/opensearch/output.go

```diff
@@ -0,0 +1,454 @@
+	"github.com/benthosdev/benthos/v4/internal/docs"
+	"github.com/benthosdev/benthos/v4/internal/http/docs/auth"
+	sess "github.com/benthosdev/benthos/v4/internal/impl/aws/session"
+	"github.com/benthosdev/benthos/v4/internal/log"
+	"github.com/benthosdev/benthos/v4/internal/message"
+	"github.com/benthosdev/benthos/v4/internal/old/util/retries"
+	itls "github.com/benthosdev/benthos/v4/internal/tls"
+)
+
+// Transport is a RoundTripper that will sign requests with AWS V4 Signing
+type Transport struct {
+	Transport http.RoundTripper
+	Creds     *credentials.Credentials
+	Signer    *v4.Signer
+	Region    string
+}
+
+func notImportedAWSOptFn(roundtripper http.RoundTripper, conf output.OpenSearchConfig) (*Transport, error) {
+	if !conf.AWS.Enabled {
+		return nil, nil
+	}
+	return nil, errors.New("unable to configure AWS authentication as this binary does not import components/aws")
+}
+
+// AWSOptFn is populated with the child `aws` package when imported.
+var AWSOptFn = notImportedAWSOptFn
+
+func init() {
+	//TODO: use public service API, refer:output_pusher.go
+	err := bundle.AllOutputs.Add(processors.WrapConstructor(func(c output.Config, nm bundle.NewManagement) (output.Streamed, error) {
+		return NewOpenSearch(c.OpenSearch, nm)
+	}), docs.ComponentSpec{
+		Name: "opensearch",
+		Summary: `
+Publishes messages into an OpenSearch index. If the index does not exist then
+it is created with a dynamic mapping.`,
+		Description: output.Description(true, true, `
+Both the `+"`id` and `index`"+` fields can be dynamically set using function
+interpolations described [here](/docs/configuration/interpolation#bloblang-queries). When
+sending batched messages these interpolations are performed per message part.
+
+### AWS
+
+It's possible to enable AWS connectivity with this output using the `+"`aws`"+`
+fields. However, you may need to set `+"`sniff` and `healthcheck`"+` to
+false for connections to succeed.`),
+		Config: docs.FieldComponent().WithChildren(
+			docs.FieldString("urls", "A list of URLs to connect to. If an item of the list contains commas it will be expanded into multiple URLs.", []string{"http://localhost:9200"}).Array(),
+			docs.FieldString("index", "The index to place messages.").IsInterpolated(),
+			docs.FieldString("action", "The action to take on the document. This field must resolve to one of the following action types: `create`, `index`, `update`, `upsert` or `delete`.").IsInterpolated().Advanced(),
+			docs.FieldString("pipeline", "An optional pipeline id to preprocess incoming documents.").IsInterpolated().Advanced(),
+			docs.FieldString("id", "The ID for indexed messages. Interpolation should be used in order to create a unique ID for each message.").IsInterpolated(),
+			docs.FieldString("type", "The document mapping type. This field is required for versions of elasticsearch earlier than 6.0.0, but are invalid for versions 7.0.0 or later.").Optional().IsInterpolated(),
```

Comment:
Should this still reference elasticsearch?

Reply:
I'll test this out. Because OpenSearch 1.x has default type as _doc and in 2.x, it is fully removed.